source/common/network/filter_manager_impl.cc

```diff
@@ -162,22 +159,22 @@ void FilterManagerImpl::onConnectionClose(ConnectionCloseAction close_action) {
     }
   }
 
-  latched_close_action_ = close_action;
-  state_.pending_local_close_ = close_action.isLocalClose();
-  state_.pending_remote_close_ = close_action.isRemoteClose();
-
-  // Only finalize if we have no pending filters.
-  // TODO(botengyao) this can be more intelligent to distinguish remote close and local
-  // close but will be more complicated.
-  if (state_.pending_close_read_filter_ == 0 && state_.pending_close_write_filter_ == 0) {
-    finalizeClose(close_action);
+  // Close the connection if we have no pending filters.
+  // The close is initialized by the connection and no filters are gating the close.
+  if (state_.filter_pending_close_count_ == 0) {
+    state_.local_close_pending_ = false;
+    state_.remote_close_pending_ = false;
+    connection_.closeConnection(close_action);
     return;
   }
 
+  latched_close_action_ = close_action;
```

Comment:
what stats do we have for latched closes that are pending (if any?)

Reply:
good point, I think it is nice to add some stats here for observability, will consider a follow up PR.